README.md
civic-client
============
Developing for CIViC involves setting up a development environment. To get started quickly, we recommend launching an AWS EC2 instance from our pre-configured and maintained AMI ([getting started wiki page](https://github.com/genome/civic-server/wiki/Getting-Started-with-CIViC-Development-on-AWS)). Alternatively, you may set up your own local development environment using the following setup instructions.
## To Install

civic-client uses npm for development, build, and resource server tasks, so ensure that you have npm (and node) installed and running. Checkout this repository and cd to it, then:

```bash
npm install
```

Then head to Starbucks, get a coffee, read that new Pikettey book everyone's been talking about, return to your workstation and install the bower package management system:

```bash
npm install -g bower
```

and finally, use bower to install civic-client's runtime libraries:

```bash
bower install
```

If bower becomes confused about which Angular version to use, pick the first one that offers version 1.3.15.

## Development

Fire up the civic-server on port 3000 and run:

```bash
gulp serve
```

Then head to `http://127.0.0.1:3001` in your browser. *NOTE:* use `127.0.0.1`, not `localhost`. A couple of the OAuth providers we use for authentication will only work if the app is loaded from `127.0.0.1`. 

The `serve` tasks starts a static file server, and a proxy that routes calls to /api to the civic-server listening on port 3000. It serves the AngularJS application, and a starts a watch task which watches all files for changes and lints, builds and injects them into the index.html accordingly.

## Production ready build - a.k.a. dist

To make the app ready for deploy to production run:

```bash
gulp build
```

The build task creates a `./dist` folder with all scripts and stylesheets concatenated, minified, and versioned, also third party libraries installed with bower will be concatenated and minified into `vendors.min.js` and `vendors.min.css` respectively.

To test the build version of the app, execute:

```bash
gulp serve:dist
```

This task executes a build, then serves the /dist directory from the same port as the `gulp serve` task, http://127.0.0.1:3001/

## Pulling Updates
As the civic-client is under heavy development, we'll be pushing releases to the master branch at a fairly rapid rate. Often, we'll update various packages and modules that are part of the workflow and/or production codebase. So after you do a `git pull` to update your local repository, be sure to:

```bash
npm install
bower install
```

This will install any new packages or modules that the new updates require.

## Related resources
* [Personalized Cancer Therapy Knowledge Base for Precision Oncology (MD Anderson Cancer Center)](https://pct.mdanderson.org/)
* [My Cancer Genome - Genetically Informed Cancer Medicine (Vanderbilt-Ingram Cancer Center)](http://www.mycancergenome.org/)
* [Targeted Cancer Care (Massachusetts General Hospital)](https://targetedcancercare.massgeneral.org/My-Trial-Guide/Mutations.aspx)

## Screenshots
**Homepage:**
![Homepage](notes/screenshots/home-view.png)

**Browse Genes in the CIViC database:**
![Homepage](notes/screenshots/gene-browse-view.png)

**View an Evidence Item:**
![Homepage](notes/screenshots/gene-variant-evidence-view.png)

**Edit a Gene:**
![Homepage](notes/screenshots/gene-edit-view.png)

**View and comment on revisions to Genes, Variants, Evidence Items:**
![Homepage](notes/screenshots/revision-view.png)

**Catch up with recent CIViC activity:**
![Homepage](notes/screenshots/activity-view.png)

**Community leaderboards and user directory:**
![Homepage](notes/screenshots/community-view.png)

### Git repositories related to the CIViC project
The CIViC source code and application are organized in a client-server model. The backend code is available in the [civic-server repository](https://github.com/genome/civic-server) and frontend code is available in the [civic-client repository](https://github.com/genome/civic-client). Issues relating to curation are tracked in the [civic-curation repository](https://github.com/genome/civic-curation). An example of a Python client is available in the [civic-api-client repository](https://github.com/griffithlab/civic-api-client). Issues relating to public CIViC meetings are tracked in the [civic-meeting repository](https://github.com/genome/civic-meeting).



CHANGELOG.md
0.0.12 / 2017-06-29
===================
* rejecting a source suggestion requires curator to select from 1 of 3 possible choices
* 'source suggestion exists' notice after submitting an identical source suggestion includes link to previosuly suggested source
* misc bug fixes

0.0.11 / 2017-05-04
===================
* added flag feature, allowing users to flag individual entities for urgent curator attention
* switching to HTML5 mode, eliminating '#' in URLs to allow better search engine indexing 
* upgraded all ui-bootstrap components to latest version

0.0.10 / 2017-02-02
===================
* Add Evidence can now pre-populate Evidence Level based on URL query string
* Evidence Rating, Evidence Summary, Evidence Type now required fields
* Filtered strings now bolded in most data grid columns
* Stats charts now adapt to responsive breakpoints
* Disease typeahead in Add Evidence form now shows disease aliases

0.0.9 / 2017/01-30
==================
* new user card design
* gene typeahead on Add Evidence form now shows gene aliases
* added new simplified overview infographic to Presentation Graphics section


0.0.8 / 2017-01-19
==================
* added Evidence, Drugs, Diseases, and Sources statistics pages
* redesigned variant menu to accomodate larger collections of variants/groups
* add evidence form can now pre-populate Clinical Significance, Evidence Type, and Evidence Direction


0.0.7 / 2017-01-12
==================
* improved documentation with new 'Getting Started' section, figure updates, more extensive documentation
* added Statistics page showing various Evidence metrics
* Source summary pages now provide a comments thread
* ConfigService now contains all attribute help strings for add/edit forms
* signout now returns users to home page, mooting various sign-out UI state issues


0.0.6 / 2016-09-29
==================
* variants now show aliases and HGVS expressions
* variant edit form provides multi-inputs for HGVS expressions and aliases
* gene advanced search now provides Suggested Revisions attribute
* Evidence Type fields on add/edit evidence item forms now provide a Predisposing type
* gene grid variant count column now filterable and sortable
* meetings page updates for NKI
* added custom header to client requests so server can better differentiate between client and direct API access
* variant name on add evidence for now a required field.
* gene variant sidebar now shows all variants instead just the first 25 returned from query


0.0.5 / 2016-08-26
==================
* sources now have their own summary view
* advanced search now provides an option to search sources
* browse view now provides option to browse sources
* data releases now shown in a table instead of long list
* evidence summary now shows citation (linked to source summary view) and pubmed ID (linked to pubmed detail page)


0.0.4 / 2016-08-18
==================
* added publication year to evidence search query builder


0.0.3 / 2016-08-11
==================
* browse view tabs no longer wrap when displayed in sm and md window widths
* user profile links now work even if user has no events returned from Search service
* users viewing their profile are now provided a button linking to the account profile edit page
* comment help text now details @admins and @editors mention notifications


0.0.2 / 2016-07-28
==================
* users may no longer approving/accept their own items/updates - approve/accept buttons will show as disabled
* HOTFIX: some components that compare entity owner to current user threw an error when rendering if user logged out
* evidence item advanced search now includes a Variant Origin option
* added note regarding CIViC partner relationships to About page
* evidence item / disease search option now provides typeahead assistance
* advanced search query builder now indicates if no records returned
* added biography attribute to profile display and edit views
* evidence grid row now indicates if pending changes exist using indicator/tooltip in ID cell


0.0.1 / 2016-07-21
==================
* gene advanced search added
* added variant count column to browse variant group grid
* added note on profile instructing users how to change their avatar on gravatar.com
* fixed variant grid variant list cell
* all entity grids can now open record in new tab if meta key (alt/command) held while clicking
* contributors no longer able to accept their own evidence items and revisions
* facebook and linkedin profile attributes now accept nicknames/handles instead of full URLs


gulpfile.js
/* jshint node: true */
'use strict';

var gulp = require('gulp')
  ,g = require('gulp-load-plugins')({lazy: false, camelize: true})
  ,noop = g.util.noop
  ,es = require('event-stream')
  ,bowerFiles = require('main-bower-files')
  ,rimraf = require('rimraf')
  ,queue = require('streamqueue')
  ,lazypipe = require('lazypipe')
  ,stylish = require('jshint-stylish')
  ,bower = require('./bower')
  ,url = require('url')
  ,proxy = require('proxy-middleware');

var isWatching = false;

var htmlminOpts = {
  removeComments: true,
  collapseWhitespace: true,
  removeEmptyAttributes: false,
  collapseBooleanAttributes: true,
  removeRedundantAttributes: true
};

/**
 * JS Hint
 */
gulp.task('jshint', function () {
  return gulp.src([
    './gulpfile.js',
    './src/app/**/*.js'
  ])
    .pipe(g.cached('jshint'))
    .pipe(jshint('./.jshintrc'))
    .pipe(livereload());
});

/**
 * CSS
 */
gulp.task('clean-css', function (done) {
  rimraf('./.tmp/css', done);
});

gulp.task('styles', ['clean-css'], function () {
  return gulp.src([
    './src/app/**/*.less',
    '!./src/app/**/_*.less'
  ])
    .pipe(g.less())
    .pipe(gulp.dest('./.tmp/css/'))
    .pipe(g.cached('built-css'))
    .pipe(livereload());
});

gulp.task('styles-dist', ['styles'], function () {
  return cssFiles().pipe(dist('css', bower.name));
});

gulp.task('csslint', ['styles'], function () {
  return cssFiles()
    .pipe(g.cached('csslint'))
    .pipe(g.csslint('./.csslintrc'))
    .pipe(g.csslint.reporter());
});

/**
 * Scripts
 */
gulp.task('scripts-dist', ['templates-dist'], function () {
  return appFiles().pipe(dist('js', bower.name, {ngAnnotate: true}));
});

/**
 * Templates
 */
gulp.task('templates', function () {
  return templateFiles().pipe(buildTemplates());
});

gulp.task('templates-dist', function () {
  return templateFiles({min: true}).pipe(buildTemplates());
});

/**
 * Vendors
 */
gulp.task('vendors', function () {
  var bowerStream = gulp.src(bowerFiles());
  return es.merge(
    bowerStream.pipe(g.filter('**/*.css')).pipe(dist('css', 'vendors')),
    bowerStream.pipe(g.filter('**/*.js')).pipe(dist('js', 'vendors'))

  );
});

/**
 * Index
 */
gulp.task('index', index);
gulp.task('build-all', ['styles', 'templates'], index);

function index () {
  var opt = {read: false};
  return gulp.src('./src/app/index.html')
    .pipe(g.inject(gulp.src(bowerFiles(), opt).pipe(stripBsDeps()), {ignorePath: 'bower_components', starttag: '<!-- inject:vendor:{{ext}} -->'}))
    .pipe(g.inject(es.merge(appFiles(), cssFiles(opt)), { ignorePath: ['.tmp', 'src/app'] }))
    .pipe(gulp.dest('./src/app/'))
    .pipe(g.embedlr())
    .pipe(gulp.dest('./.tmp/'))
    .pipe(livereload());
}

/**
 * Assets
 */
gulp.task('assets', function () {
  return gulp.src('./src/app/assets/**')
    .pipe(gulp.dest('./dist/assets'));
});

/**
 * Dist
 */
gulp.task('dist', ['vendors', 'assets', 'styles-dist', 'scripts-dist'], function () {
  return gulp.src('./src/app/index.html')
    .pipe(g.inject(gulp.src('./dist/vendors.min.{js,css}'), {ignorePath: 'dist', starttag: '<!-- inject:vendor:{{ext}} -->'}))
    .pipe(g.inject(gulp.src('./dist/' + bower.name + '.min.{js,css}'), {ignorePath: 'dist'}))
    .pipe(g.htmlmin(htmlminOpts))
    .pipe(gulp.dest('./dist/'));
});

/**
 * Static file server
 */
gulp.task('statics', function() {
    g.connect.server({
      root: ['./.tmp', './.tmp/src/app', './src/app', './bower_components']
      ,port: 3001
      ,middleware: function() {
        return [ (function() {
          var options = url.parse('http://localhost:3000');
          options.route = '/api';
          return proxy(options);
        })() ];
      }
    });
  }
);

/**
 * /dist file server
 */

gulp.task('serve-dist', function() {
    g.connect.server({
      root: ['./dist']
      ,port: 3002
      ,middleware: function() {
        return [ (function() {
          var options = url.parse('http://localhost:3000');
          options.route = '/api';
          return proxy(options);
        })() ];
      }
    });
  }
);

/**
 * Watch
 */
gulp.task('serve', ['watch']);
gulp.task('watch', ['statics', 'default'], function () {
  isWatching = true;
  // Initiate livereload server:
  g.livereload.listen();
  gulp.watch('./src/app/**/*.js', ['jshint']).on('change', function (evt) {
    if (evt.type !== 'changed') {
      gulp.start('index');
    }
  });
  gulp.watch('./src/app/index.html', ['index']);
  gulp.watch(['./src/app/**/*.html', '!./src/app/index.html'], ['templates']);
  gulp.watch(['./src/app/**/*.less'], ['csslint']).on('change', function (evt) {
    if (evt.type !== 'changed') {
      gulp.start('index');
    }
  });
});

/**
 * Default task
 */
gulp.task('default', ['lint', 'build-all']);

/**
 * Lint everything
 */
gulp.task('lint', ['jshint', 'csslint']);

/**
 * Test
 */
gulp.task('test', ['templates'], function () {
  return testFiles()
    .pipe(g.karma({
      configFile: 'karma.conf.js',
      action: 'run'
    }));
});

/**
 * Inject all files for tests into karma.conf.js
 * to be able to run `karma` without gulp.
 */
gulp.task('karma-conf', ['templates'], function () {
  return gulp.src('./karma.conf.js')
    .pipe(g.inject(testFiles(), {
      starttag: 'files: [',
      endtag: ']',
      addRootSlash: false,
      transform: function (filepath, file, i, length) {
        return '  \'' + filepath + '\'' + (i + 1 < length ? ',' : '');
      }
    }))
    .pipe(gulp.dest('./'));
});

/**
 * Test files
 */
function testFiles() {
  return new queue({objectMode: true})
    .queue(gulp.src(bowerFiles()).pipe(g.filter('**/*.js').pipe(stripBsDeps())))
    .queue(gulp.src('./bower_components/angular-mocks/angular-mocks.js'))
    .queue(appFiles())
    .queue(gulp.src(['./src/app/**/*_test.js', './.tmp/src/app/**/*_test.js']))
    .done();
}

/**
 * All CSS files as a stream
 */
function cssFiles (opt) {
  return gulp.src('./.tmp/css/**/*.css', opt);
}

/**
 * All AngularJS application files as a stream
 */
function appFiles () {
  var files = [
      './.tmp/' + bower.name + '-templates.js',
//    './.tmp/src/app/**/*.js',
//    '!./.tmp/src/app/**/*_test.js',
    './src/app/**/*.js',
    '!./src/app/**/*_test.js'
  ];
  return gulp.src(files)
    .pipe(g.angularFilesort());
}

/**
 * All AngularJS templates/partials as a stream
 */
function templateFiles (opt) {
  return gulp.src(['./src/app/**/*.html', '!./src/app/index.html'], opt)
    .pipe(opt && opt.min ? g.htmlmin(htmlminOpts) : noop());
}

/**
 * Build AngularJS templates/partials
 */
function buildTemplates () {
  return lazypipe()
    .pipe(g.ngHtml2js, {
      moduleName: bower.name + '-templates',
      prefix: '/' + bower.name + '/',
      stripPrefix: '/src/app'
    })
    .pipe(g.concat, bower.name + '-templates.js')
    .pipe(gulp.dest, './.tmp')
    .pipe(livereload)();
}

/**
 * Concat, rename, minify
 *
 * @param {String} ext
 * @param {String} name
 * @param {Object} opt
 */
function dist (ext, name, opt) {
  opt = opt || {};
  return lazypipe()
    .pipe(g.concat, name + '.' + ext)
    .pipe(gulp.dest, './dist')
    .pipe(opt.ngAnnotate ? g.ngAnnotate : noop)
    .pipe(opt.ngAnnotate ? g.rename : noop, name + '.annotated.' + ext)
    .pipe(opt.ngAnnotate ? gulp.dest : noop, './dist')
    .pipe(ext === 'js' ? g.uglify : g.minifyCss)
    .pipe(g.rename, name + '.min.' + ext)
    .pipe(gulp.dest, './dist')();
}

/**
 * Livereload (or noop if not run by watch)
 */
function livereload () {
  return lazypipe()
    .pipe(isWatching ? g.livereload : noop)();
}

/**
 * Jshint with stylish reporter
 */
function jshint (jshintfile) {
  return lazypipe()
    .pipe(g.jshint, jshintfile)
    .pipe(g.jshint.reporter, stylish)();
}

/**
 * FILTERS
 */

// we only need Bootstrap's styles, none of its JS deps
function stripBsDeps() {
  return g.filter(['*', '!**/bootstrap.js', '!**/jquery.js']);
}
